### PR TITLE
EventReceiverFirehoseMonitor

### PR DESCRIPTION
**Pull request by guobingkun**
add an EventReceiverFirehoseMonitor so that we can monitor how many
events have been queued in the EventReceiverFirehose and get a sense
about whether the firehose is under too much pressure.

### CODE REVIEW

File: server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseMonitor.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      final EventReceiverFirehoseMetric metric = entry.getValue();
+
+      final ServiceMetricEvent.Builder builder = ServiceMetricEvent.builder()
+                                                                   .setDimension("serviceName", serviceName)
+                                                                   .setDimension(
+                                                                       "bufferCapacity",
+                                                                       String.valueOf(metric.getCapacity())
+                                                                   );
+
+      emitter.emit(builder.build("ingest/firehose/bufferSize", metric.getCurrentBufferSize()));
```
Comment (fjy) [inline server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseMonitor.java]: can we please call this ingest/events/buffered

also please add this to docs

Comment (guobingkun) [inline server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseMonitor.java]: Sure. Changed metric name and added into metric doc.

File: docs/content/operations/metrics.md [hunk 2]
```diff
@@ -87,6 +87,7 @@ Memcached client metrics are reported as per the following. These metrics come d
 |`ingest/events/thrownAway`|Number of events rejected because they are outside the windowPeriod.|dataSource.|0|
 |`ingest/events/unparseable`|Number of events rejected because the events are unparseable.|dataSource.|0|
 |`ingest/events/processed`|Number of events successfully processed.|dataSource.|Equal to your # of events.|
+|`ingest/events/buffered`|Number of events queued in the EventReceiverFirehose's buffer|serviceName, bufferCapacity.|Equal to current # of events in the buffer queue.|
```
Comment (himanshug) [inline docs/content/operations/metrics.md]: I believe this metric will not get published by default, user would have to add that to druid.monitor.monitors property ?

Comment (guobingkun) [inline docs/content/operations/metrics.md]: Yeah, you are right. Modified doc.

**Comment by himanshug** [comment]
LGTM

**Comment by nishantmonu51** [comment]
Is it possible to have a generic FirehoseMonitor which can be configured in runtime.properties and emits events related to the firehose in use e.g buffered/events, batch size etc for eventReceiverFirehose and   for kafka firehose emit kafka related metrics like kafka backlog etc.
for now it can only support EventReceiverFirehose.

**Comment by cheddar** [comment]
@nishantmonu51  The metrics being exposed here are pretty specific to the `EventReceiverFirehose` so I'm not sure what extra metrics a generic firehose would meaningfully provide?

**Comment by himanshug** [comment]
@nishantmonu51 further thoughts?

**Comment by nishantmonu51** [comment]
@cheddar, @guobingkun, by generic i meant is it possible to have the glue code for emitting metrics (like FirehoseRegister and Monitor) to be reUsable by multiple firehoses so that If we implement metrics for any other firehose we wont have to write similar things again and the logic for what metrics to be emitted can be part of the firehose itself and will be specific to them.

**Comment by nishantmonu51** [comment]
btw, I am also fine with doing things as per the current impl and restructuring code when we need to emit metrics for other firehoses too.

**Comment by guobingkun** [comment]
@nishantmonu51 It's possible to make a generic FirehoseRegister + FirehoseMonitor and let each firehose implement a FirehoseMetric. Though I prefer not to do it now since it's quite a different direction from what this PR was originally going, there will also be different concurrent issues need to be resolved if we do it in a generic way.

**Comment by nishantmonu51** [comment]
:+1:

File: server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseRegister.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  public Iterable<Map.Entry<String, EventReceiverFirehoseMetric>> getMetrics()
+  {
+    return metrics.entrySet();
+  }
+
+  public void unregister(String serviceName)
+  {
+    log.info("Unregistering EventReceiverFirehoseMetric for service [%s]", serviceName);
+    metrics.remove(serviceName);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseRegister.java]: Can you check return value and log if not found?

Comment (guobingkun) [inline server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseRegister.java]: @drcrallen Added a log message

**Comment by guobingkun** [comment]
bouncing for unrelated test failure.

File: server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseRegister.java (hunk 3 above)
Comment (guobingkun) [inline server/src/main/java/io/druid/server/metrics/EventReceiverFirehoseRegister.java]: @drcrallen Any update on this one?

**Pull request merged**